presto-main/src/main/java/com/facebook/presto/sql/planner/PlanFragmenter.java

```diff
@@ -389,6 +433,242 @@ else if (exchange.getType() == ExchangeNode.Type.REPARTITION) {
+                    expectedPartitioningMetadata.getPartitioningHandle(),
+                    expectedPartitioningMetadata.getPartitionColumns().stream()
+                            .map(columnHandles::get)
+                            .collect(toImmutableList()));
+            verify(selectedLayout.getLayout().getTablePartitioning().equals(Optional.of(expectedPartitioning)), "invalid temporary table partitioning");
+
+            TableLayoutHandle layoutHandle = selectedLayout.getLayout().getHandle();
+            Map<Symbol, ColumnHandle> assignments = outputSymbols.stream()
+                    .collect(toImmutableMap(identity(), symbol -> columnHandles.get(outputColumns.get(symbol).getName())));
+
+            return new TableScanNode(
+                    idAllocator.getNextId(),
+                    tableHandle,
+                    outputSymbols,
+                    assignments,
+                    Optional.of(layoutHandle),
+                    TupleDomain.all(),
+                    TupleDomain.all());
+        }
+
+        private TableFinishNode createTemporaryTableWrite(
+                TableHandle tableHandle,
+                Map<Symbol, ColumnMetadata> temporaryTableColumns,
+                List<Symbol> outputs,
+                List<List<Symbol>> inputs,
+                List<PlanNode> sources,
+                Map<Symbol, Expression> constantExpressions,
+                PartitioningMetadata expectedPartitioningMetadata)
+        {
+            if (!constantExpressions.isEmpty()) {
+                List<Symbol> constantSymbols = ImmutableList.copyOf(constantExpressions.keySet());
+
+                // update outputs
+                outputs = ImmutableList.<Symbol>builder()
+                        .addAll(outputs)
+                        .addAll(constantSymbols)
+                        .build();
+
+                // update inputs
+                inputs = inputs.stream()
+                        .map(input -> ImmutableList.<Symbol>builder()
+                                .addAll(input)
+                                .addAll(constantSymbols)
+                                .build())
+                        .collect(toImmutableList());
+
+                // update sources
+                sources = sources.stream()
+                        .map(source -> {
+                            Assignments.Builder assignments = Assignments.builder();
+                            assignments.putIdentities(source.getOutputSymbols());
+                            constantSymbols.forEach(symbol -> assignments.put(symbol, constantExpressions.get(symbol)));
+                            return new ProjectNode(idAllocator.getNextId(), source, assignments.build());
+                        })
+                        .collect(toImmutableList());
+            }
+
+            NewTableLayout insertLayout = metadata.getInsertLayout(session, tableHandle)
+                    // Currently it is assumed that the input data has to be always
```

Comment:
I don't understand why these comments are here? -- Looks like this part of code is irrelevant to exchange ? :)

Reply:
Discussed offline